### ISSUE TITLE
Another attempt at Dict simplification

### ISSUE THREAD

**Issue by robnagler** [radiasoft/sirepo#492]
I committed to robnagler in a new [pkcollections.Dict](https://github.com/radiasoft/pykern/blob/c28f5c80bf099b12211a04d7fee208ad904cffdf/pykern/pkcollections.py#L16) that should be compatible with any dictionary.  If you have a moment, please check it out. I uses a flexible concept of item-as-attribute accesses so that you can for all practical purposes use attributes instead of item accesses. This will change code from:

``` python
if 'numberOfMacroElectrons' not in data['models']['fluxReport']:
    data['models']['fluxReport']['numberOfMacroElectrons'] = 1
```

to:

``` python
if 'numberOfMacroElectrons' not in data.models.fluxReport:
    data.models.fluxReport.numberOfMacroElectrons= 1
```

We have a lot of code with many levels of nested accesses, and this will make that code easier to read.

Please checkout `robnagler` for some testing. It's a [minor change](https://github.com/radiasoft/sirepo/commit/51b8bc2b63a534aafc4e473a541e788672cd971c#diff-40887cde02088c25476cf3326a42dbd3L314) to turn on/off the use of Dict so if there's something failing, we can use uncomment that line.

**Comment by mrakitin** [comment]
I'll have a look at it later.

**Comment by robnagler** [comment]
No rush

**Comment by moellep** [comment]
It seems to work OK, but I couldn't get it to use the new code when running py.test. For example, I changed the line in template/elegant.py generate_lattice() from

-    for bl in data['models']['beamlines']:
+    for bl in data.models.beamlines:

and it ran fine when viewing the python source from the web browser, but when I run py.test it complains:

>       for bl in data.models.beamlines:
E       AttributeError: 'dict' object has no attribute 'models'

**Comment by robnagler** [comment]
This is a problem with `py.test` vs `python setup.py test`. You probably have another `sirepo` (or `pykern`, but I think sirepo given the error) installed in `~/.pyenv/versions`. Try this:

```sh
rm -rf ~/.pyenv/versions/2.7.10/lib/python2.7/site-packages/{sirepo}*
cd ~/src/radiasoft/pykern
pip install -e .
cd ../sirepo
pip install -e .
```

At some point pip pulled in a version of sirepo from PyPI. This is done in parallel with the existing "editable" install of sirepo. `python` automatically adds `.` (cwd), but that doesn't work with running py.test in a test directory. 

I need to see if there's some workaround to prevent this from happening.

**Comment by moellep** [comment]
I removed sirepo and pykern from site-packages but am still seeing the dict
error when running from py.test or 'python setup.py test'.

**Comment by robnagler** [comment]
Did you pop install? Are you on the right branch? Force a stack trace and
send it. Thanks

**Comment by moellep** [comment]
Yes, pip install and both pykern and sirepo on robnagler branches. And it works fine through the web server, just not pytest or setup.py test. Attached is the output from py.test with --full-trace.
[trace.txt](https://github.com/radiasoft/sirepo/files/539739/trace.txt)

**Comment by moellep** [comment]
The bug shows up with the tests because it goes directly to the API which
create dict data. The webserver serializes it to disk first, so it is a
pkcollections.Dict when it is referenced. I replaced the json.load() with
simulation_db.json_load() in elegant_lattice_importer.py and wrapped the
copy() call and the test passes. pkcollections.Dict.copy() returns a dict,
which is fine.

**Comment by robnagler** [comment]
Great, thanks for finding it. I think this is a better way to work, but let
me know if you (Maxim, too!) think otherwise.

**Comment by mrakitin** [comment]
Is it in master? I wanted to use it for implementation of #371, but it fails:
```python
........
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/simulation_db.py", line 182, in fixup_old_data
    sirepo.template.import_module(data['simulationType']).fixup_old_data(data)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/srw.py", line 414, in fixup_old_data
    if 'sizeDefinition' not in data.models.gaussianBeam:
AttributeError: 'dict' object has no attribute 'models'
```

**Comment by mrakitin** [comment]
OK, I'll check your branch. We could discuss it in details after your vacation. Have a good time!

**Comment by robnagler** [comment]
I merged into master. Let me know if you have issues.
